inc/saiprefixcompression.h

```diff
@@ -0,0 +1,305 @@
+/**
+ * Copyright (c) 2024 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiprefixcompression.h
+ *
+ * @brief   This module defines SAI prefix compression interface
+ */
+
+#if !defined (__SAIPREFIXCOMPRESSION_H_)
+#define __SAIPREFIXCOMPRESSION_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIPREFIXCOMPRESSION SAI - Prefix Compression API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_table_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Label attribute used to unique identify Table.
+     *
+     * @type char
+     * @flags CREATE_AND_SET
+     * @default ""
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_LABEL = SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Prefix Compression table default entry BINCODE.
+     * BINCODE applied to any address that does not match a prefix in the table
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_DEFAULT_ENTRY_BINCODE,
+
+    /**
+     * @brief End of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_END,
+
+    /** Custom range base value */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_END
+
+} sai_prefix_compression_table_attr_t;
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_entry_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
+
+    /**
+     * @brief Prefix Compression entry BINCODE
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_BINCODE = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
+
```

Comment:
We should have an attribute to let user specify the priority of the prefix compression entry, say

SAI_PREFIX_COMPRESSION_ENTRY_ATTR_PRIORITY

Reply:
A priority does not apply in this feature. the concept of this feature is having a longest prefix entry added. The prefix table that is created is based on the longest prefix match. When creating this table if it is designed with the concept of longest prefix entries then that longest prefix would always match. 

We usually need a priority when we deal with TCAM, if a vendor plans to implement this using TCAM then this feature does not add value and ACLs could be used instead.

Reply:
In the discussion, the idea was that PREFIX compression is generic, and that ACL table is one type of consumer of the compressed metadata.  So, the assumption is more objects will be using the SAI\_PREFIX\_COMPRESSION\_ENTRY\_ATTR\_META and it could be different from ACL meta.

We should add the PRIORITY attribute, and LPM based implementations should return NOT\_SUPPORTED in the sai\_capability\_query() for this attribute. I don't see why we should not keep it generic.

Reply:
Yes this feature is generic but generic with the use of LPM capabilities to optimize the correlation between a LPM address and a meta value.  The entries are a LPM entry. Can you give a situation where you would want a more specific entry to take place over a longest prefix entry? And why would we do that?

If the intention of the feature is not to use LPM then you could use ACLs instead to match on a specific set of entries with a specific priority. This feature is really intended to allow the consolidation of addresses within a single LPM address.